client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -157,22 +162,37 @@ private void handleMissingLog() {
      * message find which new segment it should go to and send it there. 
      */
     private void handleLogSealed(Segment segment) {
-        retransmitPool.submit(() -> {
-            writeFlushLock.readLock().lock();
-            /* Using segmentSealedLock the following behaviour is enforced
-               - Prevent concurrent segmentSealedCallback for different segments from being invoked concurrently.
-               - Ensure waiting segmentSealedCallbacks are invoked before the next write is invoked.
-               This ensures that resend() would be invoked again if we observe a segment sealed exception.
-             */
-            segmentSealedLock.lock();
-            try {
-                List<PendingEvent> toResend = selector.refreshSegmentEventWritersUponSealed(segment, segmentSealedCallBack);
-                resend(toResend);
-            } finally {
-                segmentSealedLock.unlock();
-                writeFlushLock.readLock().unlock();
-            }
-        });
+        /*
+         * Using segmentSealedLock prevents concurrent segmentSealedCallback for different segments
+         * from being invoked concurrently.
+         * 
+         * By calling flush when while the write lock is held we can ensure that any inflight
+         * entries that will succeed in being written to a new segment are written and any
+         * segmentSealedCallbacks will be be called happen before the next write is invoked.
+         */
+        writeFlushLock.readLock().lock();
+        Retry.indefinitelyWithExpBackoff(config.getInitalBackoffMillis(), config.getBackoffMultiple(),
+                                         config.getMaxBackoffMillis(),
+                                         t -> log.error("Encountered excemption when handeling a sealed segment: ", t))
+             .runAsync(() -> {
+                 // Running in the shrinking pool as this operation is infrequent and slow.
+                 return CompletableFuture.runAsync(() -> {
+                     segmentSealedLock.lock();
+                     try {
+                         resend(selector.refreshSegmentEventWritersUponSealed(segment, segmentSealedCallBack));
+                     } finally {
+                         segmentSealedLock.unlock();
+                     }
+                     /* In the case of segments merging Flush ensures there can't be anything left
+                      * inflight that will need to be resent to the new segment when the write lock
+                      * is released. (To preserve order)
+                      */
+                     flushInternal();
+                 }, retransmitPool);
+             }, backgroundThreadPool).handle((r, t) -> {
```

Comment:
Observation: writeFlushLock ( which is of type StampedLock()) is not a rentrant lock.

In this change, readLock is obtained in the client's internal thread pool and released in the handle() method of the future which is running on retransmitPool. Is it better to ensure  `writeFlushLock.readLock().lock();`  and `writeFlushLock.readLock().unlock() ` happen on the same thread?

Reply:
Yes, I switched to the stamped lock so it could be released in a different thread.